DevTools Injector Graph Visualization polishing

AleksanderBodurri opened the pull request:
Some cleanup before release.

See individual commits

AleksanderBodurri committed:
refactor(devtools): stop inspector from changing tab to components

Previously this was the only page that used the inspector, so we automatically changed the tab out of convenience.

Now, the injector tree tab also does uses the inspector for some functionality, so we disable this behaviour.

AleksanderBodurri committed:
feat(devtools): squash multi providers into 1 and allow them to be logged to the console

Implements a feature allowing users to visualize multiproviders as one row in the providers table.

Also enables the user to log to console any provider in the table. This log includes information about the provider selected, the injector it was configured in, and also calls `Injector.get(Token)` to determine the value the provider evaluates to in it's injector.

AleksanderBodurri committed:
perf(devtools): optimize injector graph discovery

Our algorithm for discovering the injector graph of an application involves calculating resolution paths for each angular node on a page, and then using those paths to construct the underlying tree.

Along the we way we perform many expensive computations that are candidates for optimization through caching. This commit implements this caching, resulting in a substantial increase in performance for large applications.

AleksanderBodurri committed:
feat(devtools): implement zoom and pan in injector graph visualizer

Previously only the trackpad could be used to navigate this view.

Now we can zoom and pan around using our mouse wheel and scroll.

Additionally, this commit fixes many issues related to the visualization of the injector graph visualization, allowing it to be more compact without impacting legibility and minimizing edge collisions in larger graphs.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/dependency-injection/injector-tree-visualizer.ts:
```diff
@@ -58,44 +67,100 @@ export abstract class GraphRenderer<T, U> {
-      private _orientation: 'horizontal'|'vertical' = 'horizontal',
-      private _nodeSize: [width: number, height: number] = [70, 200],
-      private _nodeSeperation: number = 2,
+      public config: {
```
Suggestion: It looks like the optionality of the inputs is leaking into the `config` type, requiring `config.orientation!` even though it is given a default value. If we could put defaults for properties directly in the constructor that would work, but I don't think we can given the `public` modifier, since we need to name the property. I think the better way to reassign it:

```typescript
interface InjectorTreeVisualizerConfig {
    orientation: 'horizontal'|'vertical',
    nodeSize: [width: number, height: number],
    nodeSeperation: (nodeA: InjectorTreeD3Node, nodeB: InjectorTreeD3Node) => number,
    nodeLabelSize: [width: number, height: number]
}

class InjectorTreeVisualizer {
    public config: InjectorTreeVisualizerConfig;

    constructor(/* ... */, {
        orientation = 'horizontal',
        nodeSize = [200, 500],
        nodeSeparation = () => 2,
        nodeLabelSize = [250, 60],
    }: Partial<InjectorTreeVisualizerConfig> = {}) {
        this.config = { orientation, nodeSize, nodeSeparation, nodeLabelSize };
    }
}
```

This at least constrains the optionality of the properties to the constructor, and not the `config` property.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.html:
```diff
@@ -1,33 +1,59 @@
 <p class="not-supported" *ngIf="!diDebugAPIsAvailable">
-  This feature is not supported by your version of Angular. Please update to version 17.0.0 or higher.
+  This feature only available on version 17.0.0 or higher of Angular.
```
Typo: "This feature **is** only available on..."

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -41,24 +45,31 @@ export class InjectorTreeComponent {
-      this.setUpElementInjectorVisualizer();
-    });
-  }
+  private _messageBus = inject(MessageBus<Events>);
```
Question: I've never seen a generic used in an `inject` before. I didn't even know you could put a type parameter on a class value reference. Isn't this doing an unchecked cast? I would suspect that if a module provided `MessageBus<OtherEvents>` that would be injected here without error. I wonder if `inject(MessageBus) as MessageBus<Events>` would at least be more clear that it performing a hard cast.

I might follow up with some framework folks to see if this is intended or desirable.

AleksanderBodurri replied:
Ahh good point, you're right this is being inferred as MessageBus<any> without the `as`

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -88,60 +135,118 @@ export class InjectorTreeComponent {
+      if (this.firstRender) {
+        this.snapToRoot(this.injectorTreeGraph);
+        this.snapToRoot(this.elementInjectorTreeGraph);
+      }
```
Consider: Could we maybe use [`afterNextRender`](https://angular.io/api/core/afterNextRender) in the constructor to trigger on only the first render?

I don't know if it would mess with the timing of the rest of this functionality, might not be worth it if it causes more timing problems.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -67,17 +78,53 @@ export class InjectorTreeComponent {
-              return 1;
-            }
-            if (!aIsInjectionToken && bIsInjectionToken) {
+            if (a.token < b.token) {
```
Question: Tokens are strings at this point right? Would `return a.token.localeCompare(b.token);` be more appropriate? That also allows you to drop the conditionals (for numbers you can usually do `return a - b;`, but that doesn't work with strings). I'm actually surprised this works at all with strings, I would have expected JS to convert them to numbers by length or something.

AleksanderBodurri replied:
TIL

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -67,17 +78,53 @@ export class InjectorTreeComponent {
         'latestInjectorProviders',
         (_: SerializedInjector, providers: SerializedProviderRecord[]) => {
           this.providers = providers.sort((a, b) => {
```
Nit: I can see this is preexisting, but since I noticed now `Array.prototype.sort` sorts elements _in-place_ meaning this will mutate the array provided as an input. I suggest cloning it before sorting to avoid mutating state unintentionally.

```typescript
this.providers = Array.from(providers).sort(() => /* ... */);
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#return_value

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -67,17 +78,53 @@ export class InjectorTreeComponent {
+    }
+  }
+
+  toggleHideAngularInjectors(): void {
```
Nit: `toggleHideAngularInjectors` and `toggleHideInjectorsWithNoProviders` are pretty similar. Could we factor out the shared pieces to some kind of private utility function?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -88,60 +135,118 @@ export class InjectorTreeComponent {
+      if (node.data.injector.type === 'environment') {
+        this.injectorTreeGraph.snapToNode(node);
+      }
+    }, 100);
```
Nit: If we're only waiting for CD, then we shouldn't need `100` right?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -88,60 +135,118 @@ export class InjectorTreeComponent {
+        this.elementInjectorTreeGraph.snapToNode(node);
+      }
+
+      if (node.data.injector.type === 'environment') {
```
Nit: Use an `else if` to more clearly communicate that these two paths cannot happen in the same execution.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -151,7 +256,7 @@ export class InjectorTreeComponent {
     this.injectorTreeGraph?.cleanup?.();
     this.injectorTreeGraph = new InjectorTreeVisualizer(
-        new D3GraphRenderer(this.svgContainer.nativeElement, this.g.nativeElement));
+        this.svgContainer.nativeElement, this.g.nativeElement, {nodeSize: [200, 500]});
```
Question: Isn't this the default `nodeSize`? Can we just use the default or do we care strongly enough that it _needs_ to be `[200, 500]` in this situation?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -161,7 +266,8 @@ export class InjectorTreeComponent {
     this.elementInjectorTreeGraph = new InjectorTreeVisualizer(
-        new D3GraphRenderer(this.elementSvgContainer.nativeElement, this.elementG.nativeElement));
+        this.elementSvgContainer.nativeElement, this.elementG.nativeElement,
+        {nodeSize: [200, 500], nodeSeperation: () => 1});
```
Restating the default here as well.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -219,20 +329,29 @@ export class InjectorTreeComponent {
+    nodes.forEach(node => {
+      node.classList.remove('selected');
+      node.classList.remove('highlighted');
+    });
```
Nit: Prefer native for loop.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -268,33 +387,45 @@ export function findExistingPath(
     let currentLevel = injectorTree;
 
+    let i = 0;
     for (const injector of path) {
```
Suggestion: Can we use `Object.entries` here as well?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -268,33 +387,45 @@ export function findExistingPath(
         injector: injector,
         children: [],
       });
+      const next = currentLevel[currentLevel.length - 1];
```
Question: How are we sure `currentLevel.length > 0`? Should we add a check somewhere or throw an assertion error?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -307,16 +438,16 @@ export function grabInjectorPathsFromDirectiveForest(directiveForest: DevToolsNo
     node.children.forEach(child => grabInjectorPaths(child));
   };
-  grabInjectorPaths(directiveForest[0]);
+  directiveForest.forEach(directive => grabInjectorPaths(directive));
```
Nit: Prefer native `for` loop.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -307,16 +438,16 @@ export function grabInjectorPathsFromDirectiveForest(directiveForest: DevToolsNo
-  const elementPaths: InjectorPath[] = [];
-  const environmentPaths: InjectorPath[] = [];
+  let elementPaths: InjectorPath[] = [];
+  let environmentPaths: InjectorPath[] = [];
```
Nit: I think `environmentPaths` can stay `const`?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -344,6 +475,9 @@ export function splitInjectorPathsIntoElementAndEnvironmentPaths(injectorPaths:
+  elementPaths =
+      elementPaths.filter(({path}) => path.every(injector => injector.type === 'element'));
+
   return {elementPaths, environmentPaths, startingElementToEnvironmentPath};
```
Nit: Rather than rebinding an existing variable, put the filter on the returned value.

```typescript
return {
  elementPaths: elementPaths.filter(({path}) => /* ... */),
  environmentPaths,
  startingElementToEnvironmentPath,
};
```

This also means `elementPaths` can stay `const` as well.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -396,10 +530,17 @@ const ANGULAR_DIRECTIVES = [
 ];
 
```
Follow up: How will we keep this list of directives up to date? It's somewhat rare, but we do add directives occasionally (`NgOptimizedImage` being a recent example).

I wonder if it would be better to keep this information in lists in `@angular/core` and `@angular/common` which we can import here. Something to talk with FW people about and follow up after v17.

AleksanderBodurri replied:
Agreed, I think it would be ideal to have a better way of determining what is a framework directive, provider and injector

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/injector-tree/injector-tree.component.ts:
```diff
@@ -396,10 +530,17 @@ const ANGULAR_DIRECTIVES = [
+  const paths = [...injectorPaths];
+  paths.forEach((injector) => {
+    injector.path = injector.path.filter(({providers}) => providers === undefined || providers > 0);
+  });
```
Nit: Prefer native for loops.

You could probably map over these values and update the `path` for each one, but I think that's likely to be worse than what you have already.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools.component.scss:
```diff
@@ -103,55 +107,89 @@
+  
+      .node-null {
+        border: 1px solid #8b8b8b;
+        background: #ffffff;
```
Suggestion: Prefer plain color names where possible. `white` is an alias of `#ffffff` and `black` is an alias of `#000000`.

dgp1130 reviewed (approved):
Thanks for putting this all together, looks like a ton of really great usability improvements.

I mostly left a lot of minor code quality suggestions but I don't see anything overly concerning here. Overall it looks great!

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -288,3 +288,38 @@ const getInjectorProvidersCallback = (messageBus: MessageBus<Events>) =>
       messageBus.emit('latestInjectorProviders', [injector, serializedProviderRecords]);
     };
+
+const logProvider = (injector: SerializedInjector, provider: SerializedProviderRecord): void => {
```
Ahh good catch, the message bus registration was dropped accidentally while I was grouping code changes to specific commits

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -72,23 +72,24 @@ const getLatestComponentExplorerViewCallback = (messageBus: MessageBus<Events>)
-        forest = prepareForestForSerialization(
-            initializeOrGetDirectiveForestHooks().getIndexedDirectiveForest());
       }
+      injectorsSeen.clear();
```
Totally fair, I considered wrapping all of the caching logic in an `InjectorCache` class but wanted this initial state to be as easy to change as possible

eneajaho commented:
![CleanShot 2023-11-03 at 09 05 39@2x](https://github.com/angular/angular/assets/25394362/148f7f40-d1aa-457c-a69f-548c4c435973)

The h1 title of Environment Hierarchy look really bad in a 14inch screen. 


![CleanShot 2023-11-03 at 09 09 30@2x](https://github.com/angular/angular/assets/25394362/d42743e3-0a86-4ff6-8b17-63e0a78d4cce)

Just added: 
```
.injector-hierarchy  {
    ... 
    height: 30px;
    font-size: 15px;
    padding-left: 20px;
}
```

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -274,16 +275,44 @@ const getInjectorProvidersCallback = (messageBus: MessageBus<Events>) =>
+      const serializedProviderRecords: SerializedProviderRecord[] = [];
+
+      for (const token of tokenToRecords.keys()) {
+        const records = tokenToRecords.get(token)!;
```
I think `records` needs to be pulled into the `const` declaration? Otherwise there's not much point in `entries`.

dgp1130 reviewed (approved):
Looks great! If we can get CI to pass hopefully we can merge this today then?

AleksanderBodurri committed:
test(devtools): fix transformInjectorResolutionPathsIntoTree tests

Previously transformInjectorResolutionPathsIntoTree returned an array, now it returns a tree node so we update the test cases to reflect that.

atscott commented:
This PR was merged into the repository by commit 63777d20d99ca6206420d545182a754b8f2dc2d8.
